src/main/java/com/uber/cadence/testing/WorkflowShadower.java

```diff
@@ -0,0 +1,116 @@
+import com.uber.cadence.internal.shadowing.ReplayWorkflowActivityImpl;
+import com.uber.cadence.internal.shadowing.ScanWorkflowActivity;
+import com.uber.cadence.internal.shadowing.ScanWorkflowActivityImpl;
+import com.uber.cadence.serviceclient.IWorkflowService;
+import com.uber.cadence.shadower.Mode;
+import com.uber.cadence.shadower.ReplayWorkflowActivityParams;
+import com.uber.cadence.shadower.ReplayWorkflowActivityResult;
+import com.uber.cadence.shadower.ScanWorkflowActivityParams;
+import com.uber.cadence.shadower.ScanWorkflowActivityResult;
+import com.uber.cadence.worker.ShadowingOptions;
+import com.uber.m3.tally.NoopScope;
+import com.uber.m3.tally.Scope;
+import java.time.Duration;
+import java.time.ZonedDateTime;
+
+public final class WorkflowShadower {
+
+  private final ShadowingOptions options;
+  private final String query;
+  private final ScanWorkflowActivity scanWorkflow;
+  private final ReplayWorkflowActivity replayWorkflow;
+
+  public WorkflowShadower(IWorkflowService service, ShadowingOptions options, String taskList) {
+
+    this(service, options, taskList, new NoopScope());
+  }
+
+  public WorkflowShadower(
+      IWorkflowService service, ShadowingOptions options, String taskList, Scope metricsScope) {
+    this(
+        options,
+        new ScanWorkflowActivityImpl(service),
+        new ReplayWorkflowActivityImpl(service, metricsScope, taskList));
+  }
+
+  public WorkflowShadower(
+      ShadowingOptions options,
+      ScanWorkflowActivity scanWorkflow,
+      ReplayWorkflowActivity replayWorkflow) {
+    this.options = options;
+    this.query = options.getWorkflowQuery();
+    this.scanWorkflow = scanWorkflow;
+    this.replayWorkflow = replayWorkflow;
+  }
+
+  public void run() throws Throwable {
+    byte[] nextPageToken = null;
+    int replayCount = 0;
+
+    int maxReplayCount = Integer.MAX_VALUE;
+    Duration maxReplayDuration = Duration.ZERO;
+    ZonedDateTime now = ZonedDateTime.now();
+    if (options.getExitCondition() != null) {
+      if (options.getExitCondition().getShadowCount() != 0) {
+        maxReplayCount = options.getExitCondition().getShadowCount();
+      }
+      if (options.getExitCondition().getExpirationIntervalInSeconds() != 0) {
+        maxReplayDuration =
+            Duration.ofSeconds(options.getExitCondition().getExpirationIntervalInSeconds());
+      }
+    }
+    do {
+      ScanWorkflowActivityResult ScanResult =
+          scanWorkflow.scan(
+              new ScanWorkflowActivityParams()
+                  .setDomain(options.getDomain())
+                  .setWorkflowQuery(query)
+                  .setNextPageToken(nextPageToken)
+                  .setSamplingRate(options.getSamplingRate()));
+      nextPageToken = ScanResult.getNextPageToken();
+
+      for (WorkflowExecution execution : ScanResult.getExecutions()) {
+        ReplayWorkflowActivityResult replayResult =
+            replayWorkflow.replay(
```

Comment:
Is there any logging inside the replay activity for why the replay failed or skipped?

Reply:
yes